prerequisite docs (#2147)
growthbook/growthbook: docs/docs/features/prerequisites.mdx

There are two types of prerequisite features in GrowthBook: **Top-level Prerequisites** (feature gating) and **Inline Prerequisite Targeting** (rule-level and experiment-level gating).

## Top-Level Prerequisites

Top-level prerequisites are defined per dependent feature. They function similarly to a feature's kill switches: if the prerequisite feature is not serving `true` then the dependent feature will not be enabled.

:::note
**Top-level Prerequisites** is a GrowthBook Pro and Enterprise feature.
:::

<img src="/images/features/feature-top-level-prerequisites.png" alt="Top-level prerequisites" width="700" />

Add one or more top-level prerequisites by clicking the "Add Prerequisite Feature" button on the feature page and then selecting the feature you want to use as a prerequisite.

In order for a feature to be eligible to be a top-level prerequisite, it must be a boolean (true/false). Also, it must be in the same project as the dependent feature.

### Prerequisite states and values

A summary of the prerequisite state and value will show while adding or editing a prerequisite, as well as on the dependent feature page.

> - **Deterministic** states (**live** and **not live**) are applied to your features within the GrowthBook.

Reviewer: ```suggestion
- **Deterministic** states (**live** and **not live**) are applied to your features within GrowthBook.
```

  - **Live** prerequisites which are serving `false` will still block their dependent features (`false` ≠ `true`)
  - **Not live** prerequisites will always block their dependent features (they evaluate to `null`, and `null` ≠ `true`)

:::info
Prerequisites with deterministic states work "out of the box" regardless of SDK version support.

Any feature that is always "not live" will not be seen by the SDK. Any feature that is always "live" will no longer reference its prerequisites in the SDK. This means that no SDK-level evaluation of prerequisites is needed (these prerequisites work irrespective of SDK version support).
:::

- **Schrödinger** state means that the prerequisite's state cannot be determined in advance. It may depend on user attributes or other non-deterministic factors. (This is homage to the physicist Erwin Schrödinger who proposed a thought experiment involving a cat in a box that is both alive and dead at the same time.)

:::info
Prerequisites with a Schrödinger state must be evaluated at runtime in the SDK. Prerequisite evaluation is currently supported in the following SDK versions:

- JavaScript: `0.34.0`
- React: `0.24.0`

>   :::

Reviewer: Messing up formatting
```suggestion

:::
```

## Inline Prerequisite Targeting

Inline prerequisite targeting allows finer-grained control over prerequisite behavior than top-level prerequisites.

> 1. Inline prerequisites can be applied a feature's [override rule](/features/rules) level, and can be environment-specific. This comes with the added benefits of feature draft releases and approvals.

Reviewer: ```suggestion
1. Inline prerequisites can be applied at a feature's [override rule](/features/rules) level, and can be environment-specific. This comes with the added benefits of feature draft releases and approvals.
```

2. Inline prerequisites can be applied to individual experiments which may not be linked to a specific feature (such as visual experiments).
3. Inline prerequisite targeting is not limited to boolean features. You can specify any evaluation condition you'd like (ex: prerequisite value is: greater than 3, in a list of allowed values, or matches a regex pattern). You can even do advanced targeting with JSON.

:::note
**Inline Prerequisite Targeting** is a GrowthBook Enterprise feature only.

> :::

Reviewer: Formatting being messed up.  I think an extra newline would help?
```suggestion

:::
```

<img src="/images/features/feature-inline-prerequisite-targeting.png" alt="Inline prerequisite targeting" width="700" />

### Inline prerequisite states and values

The same **deterministic** and **Schrödinger** states apply to inline prerequisites as they do to top-level prerequisites ([see above](#prerequisite-states-and-values)). Below is a summary of how they apply to inline prerequisites:

- **Deterministic** states (**live** and **not live**) are calculated using your evaluation condition, which is **_not_** limited to `is true`. As before, no run-time evaluation of prerequisites is required in the SDK.

  - **Live** prerequisites which pass the evaluation condition will allow their dependent rules or experiments to be enabled.
  - **Live** prerequisites which fail the evaluation condition will still block their dependent rules or experiments.

>   - **Not live** prerequisites will always block their dependent rules or experiments.

Reviewer: This isn't quite true

```suggestion
  - **Not live** prerequisites will generally block the dependent rule or experiment, unless the evaluation condition specifically checks for this (e.g. `is not live`)
```
